File: Course 4/Modern Programming Technologies/MPT/MPT/TFrac.cpp
#include "TFrac.h"

TFrac::TFrac() 
{
	num = 0;
	den = 1;
}

TFrac::TFrac(int a, int b) : num(a), den(b)
{
	if (den == 0)
	{
		std::cerr << "Denominator can't be 0!" << std::endl;
		num = 0;
		den = 1;
		return;
	}

	int gcd = std::gcd(num, den);
	num /= gcd;
	den /= gcd;
	if (den < 0)
	{
		num *= -1;
		den *= -1;
	}
}

TFrac::TFrac(std::string frac)
{
	std::string s_num, s_den;
	size_t slash = frac.find('/');
	s_num = frac.substr(0, slash);
	s_den = frac.substr(slash + 1, frac.length());
	num = std::stoi(s_num);
	den = std::stoi(s_den);

	if (den == 0)
	{
		std::cerr << "Denominator can't be 0!" << std::endl;
		num = 0;
		den = 1;
		return;
	}

	int gcd = std::gcd(num, den);
	num /= gcd;
	den /= gcd;
	if (den < 0)
	{
		num *= -1;
		den *= -1;
	}
}

TFrac TFrac::copy()
{
	return {num, den};
}

TFrac TFrac::add(TFrac& frac)
{
	TFrac res;
	res.num = num * frac.den + frac.num * den;
	res.den = den * frac.den;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}


TFrac TFrac::operator+(TFrac& frac)
{
	TFrac res;
	res.num = num * frac.den + frac.num * den;
	res.den = den * frac.den;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}

TFrac TFrac::multiply(TFrac& frac)
{
	TFrac res;
	res.num = num * frac.num;
	res.den = den * frac.den;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}

TFrac TFrac::operator*(TFrac& frac)
{
	TFrac res;
	res.num = num * frac.num;
	res.den = den * frac.den;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}


TFrac TFrac::subtract(TFrac& frac)
{
	TFrac res;
	res.num = num * frac.den - frac.num * den;
	res.den = frac.den * den;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}

TFrac TFrac::operator-(TFrac& frac)
{
	TFrac res;
	res.num = num * frac.den - frac.num * den;
	res.den = frac.den * den;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}

TFrac TFrac::divide(TFrac& frac)
{
	if (frac.num == 0 || den == 0)
	{
		std::cout << "No divinition by 0!" << std::endl;
		return { 0, 1 };
	}
	TFrac res;
	res.num = num * frac.den;
	res.den = den * frac.num;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}

TFrac TFrac::operator/(TFrac& frac)
{
	if (frac.num == 0 || den == 0)
	{
		std::cout << "No divinition by 0!" << std::endl;
		return { 0, 1 };
	}
	TFrac res;
	res.num = num * frac.den;
	res.den = den * frac.num;
	int gcd = std::gcd(res.num, res.den);
	res.num /= gcd;
	res.den /= gcd;
	if (res.den < 0)
	{
		res.num *= -1;
		res.den *= -1;
	}
	return res;
}

TFrac TFrac::square()
{
	return { num * num, den * den };
}

TFrac TFrac::reverse()
{
	return { den, num };
}


TFrac TFrac::operator-()
{
	return { -num, den };
}

bool TFrac::operator==(TFrac frac)
{
	return num == frac.num && den == frac.den;
}

bool TFrac::operator>(TFrac frac)
{
	return num * frac.den > frac.num * den;
}

int TFrac::getNumeratorNum()
{
	return num;
}

int TFrac::getDenominatorNum()
{
	return den;
}

std::string TFrac::getNumeratorString()
{
	return std::to_string(num);
}

std::string TFrac::getDenominatorString()
{
	return std::to_string(den);
}

std::string TFrac::getFractionString()
{
	return std::to_string(num) + '/' + std::to_string(den);
}

File: Course 4/Modern Programming Technologies 2/MPT-RGZ/MPT-RGZ/TFrac.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace MPT_RGZ
{
    public class TFrac
    {
        private long numerator;
        public long Numerator { get; }
        private long denominator;
        public long Denominator { get; }

        static void Swap<T>(ref T lhs, ref T rhs)
        {
            T temp;
            temp = lhs;
            lhs = rhs;
            rhs = temp;
        }

        public static long GCD(long a, long b)
        {
            a = Math.Abs(a);
            b = Math.Abs(b);
            while (b > 0)
            {
                a %= b;
                Swap(ref a, ref b);
            }
            return a;
        }

        public TFrac()
        {
            numerator = 0;
            denominator = 1;
        }

        public TFrac(long a, long b)
        {
            if (a < 0 && b < 0)
            {
                a *= -1;
                b *= -1;
            }
            else if (b < 0 && a > 0)
            {
                b *= -1;
                a *= -1;
            }
            else if (a == 0 && b == 0 || b == 0 || a == 0 && b == 1)
            {
                numerator = 0;
                denominator = 1;
                return;
            }
            numerator = a;
            denominator = b;
            long gcdRes = GCD(a, b);
            if (gcdRes > 1)
            {
                numerator /= gcdRes;
                denominator /= gcdRes;
            }
        }

        public TFrac(string frac)
        {
            Regex FracRegex = new Regex(@"^-?(\d+)/(\d+)$");
            Regex NumberRegex = new Regex(@"^-?\d+/?$");
            if (FracRegex.IsMatch(frac))
            {
                List<string> FracParts = frac.Split('/').ToList();
                numerator = Convert.ToInt64(FracParts[0]);
                denominator = Convert.ToInt64(FracParts[1]);
                if (denominator == 0)
                {
                    numerator = 0;
                    denominator = 1;
                    return;
                }
                long gcdResult = GCD(numerator, denominator);
                if (gcdResult > 1)
                {
                    numerator /= gcdResult;
                    denominator /= gcdResult;
                }
                return;
            }
            else if (NumberRegex.IsMatch(frac))
            {
                if (long.TryParse(frac, out long NewNumber))
                    numerator = NewNumber;
                else
                    numerator = 0;
                denominator = 1;
                return;
            }
            else
            {
                numerator = 0;
                denominator = 1;
                return;
            }
        }

        public TFrac(TFrac anotherFrac)
        {
            numerator = anotherFrac.numerator;
            denominator = anotherFrac.denominator;
        }

        public void SetString(string str)
        {
            TFrac TempFrac = new TFrac(str);
            numerator = TempFrac.numerator;
            denominator = TempFrac.denominator;
        }

        public TFrac Add(TFrac a)
        => new TFrac(numerator * a.denominator + denominator * a.numerator, denominator * a.denominator);

        public TFrac Mul(TFrac a)
        => new TFrac(numerator * a.numerator, denominator * a.denominator);

        public TFrac Sub(TFrac a)
        => new TFrac(numerator * a.denominator - denominator * a.numerator, denominator * a.denominator);

        public TFrac Div(TFrac a)
        => new TFrac(numerator * a.denominator, denominator * a.numerator);

        public TFrac Square()
        => new TFrac(numerator * numerator, denominator * denominator);

        public TFrac Reverse() => new TFrac(denominator, numerator);

        public TFrac Minus() => new TFrac(-numerator, denominator);

        public bool Equal(TFrac a) => numerator == a.numerator && denominator == a.denominator;

        public static bool operator >(TFrac a, TFrac b) => (Convert.ToDouble(a.numerator) / Convert.ToDouble(a.denominator)) > (Convert.ToDouble(b.numerator) / Convert.ToDouble(b.denominator));

        public static bool operator <(TFrac a, TFrac b) => (Convert.ToDouble(a.numerator) / Convert.ToDouble(a.denominator)) < (Convert.ToDouble(b.numerator) / Convert.ToDouble(b.denominator));

        public long getNumeratorNum() => numerator;

        public long getDenominatorNum() => denominator;

        public string getNumeratorString() => numerator.ToString();

        public string getDenominatorString() => denominator.ToString();

        public override string ToString() => getNumeratorString() + "/" + getDenominatorString();
    }
}
